saged/normalize_recount_data.py

```diff
@@ -0,0 +1,178 @@
+import tqdm
+
+
+def parse_gene_lengths(file_path: str) -> Dict[str, int]:
+    """Parses a tsv file containing genes and their length
+
+    Arguments
+    ---------
+    file_path - The path to the file mapping genes to lengths
+
+    Returns
+    -------
+    gene_to_len - A dict mapping ensembl gene ids to their lenght in base pairs
+    """
+    gene_to_len = {}
+    with open(file_path) as in_file:
+        # Throw out header
+        in_file.readline()
+        for line in in_file:
+            line = line.replace('"', '')
+            gene, length = line.strip().split('\t')
+            try:
+                gene_to_len[gene] = int(length)
+            except ValueError:
+                # Some genes have no length, but will be removed in a later step
+                pass
+    return gene_to_len
+
+
+def calculate_tpm(counts: np.ndarray, gene_length_arr: np.ndarray) -> np.ndarray:
+    """"Given an array of counts, calculate the transcripts per kilobase million
+
+    Arguments
+    ---------
+    counts: The array of transcript counts per gene
+    gene_length_arr: The array of lengths for each gene in counts
+
+    Returns
+    -------
+    tpm: The tpm normalized expression data
+    """
+    counts = np.array(counts, dtype=float)
+
+    reads_per_kb = counts / gene_length_arr
+
+    sample_total_counts = np.sum(counts)
+    per_million_transcripts = sample_total_counts / 1e6
+
+    tpm = reads_per_kb / per_million_transcripts
+
+    return tpm
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument('count_file', help='The file containing the count matrix generated by '
+                                           'download_recount3.R')
+    parser.add_argument('gene_file', help='The file with gene lengths from get_gene_lengths.R')
+    parser.add_argument('out_file', help='The file to save the normalized results to')
+    parser.add_argument('--num_genes', help='The script will keep the top K most variable genes. '
+                                            'This flag sets K.', default=5000, type=int)
+
+    args = parser.parse_args()
+
+    gene_to_len = parse_gene_lengths(args.gene_file)
+
+    with open(args.count_file, 'r') as count_file:
+        header = count_file.readline()
+        header = header.replace('"', '')
+        header_genes = header.strip().split('\t')
+        header_genes = [gene.split('.')[0] for gene in header_genes]
+
+        bad_indices = []
+        gene_length_arr = []
+        for i, gene in enumerate(header_genes):
+            if gene not in gene_to_len.keys():
+                bad_indices.append(i)
+            else:
+                gene_length_arr.append(gene_to_len[gene])
+
+        gene_length_arr = np.array(gene_length_arr)
+
+        means = None
+        M2 = None
+        maximums = None
+        minimums = None
+
+        # First time through the data, calculate statistics
+        for i, line in tqdm.tqdm(enumerate(count_file), total=317475):
```

Comment:
What is this `total=317475` number? Is that liable to change if recount changes at all?

Reply:
317475 is the number of lines in the file. Tqdm generates a progress bar for the loop and estimates the time remaining. Since reading all the way through the file to calculate length is slow, I just gave it the number of lines manually. This number will change with recount3, but the bug that occurs will be a slight misestimation of how long the loop will take.

To avoid using magic numbers I'll add a constant explaining what that number is